daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala

```diff
@@ -0,0 +1,328 @@
+      val newPkg1 = templatePkg.copy(languageVersion = newVersion)
+      val newPkg2 = interfacesPkg.copy(languageVersion = newVersion)
+      val pkgs = SpeedyTestLib.typeAndCompile(
+        Map(
+          templatePkgId -> newPkg1,
+          interfacesPkgId -> newPkg2,
+          implementsPkgId -> oldPkg1,
+          coImplementsPkgId -> oldPkg2,
+        )
+      )
+
+      for ((templateId, interfaceId, contract) <- testData) {
+        val result = evaluateBeginExercise(
+          pkgs,
+          templateId,
+          Some(interfaceId),
+          contractId,
+          committers = Set(contractParty),
+          controllers = Set(contractParty),
+          getContract = Map(contractId -> contract),
+        )
+
+        inside(result) { case Right(transaction) =>
+          transaction.version shouldBe TransactionVersion.assignNodeVersion(newVersion)
+        }
+      }
+    }
+
+    "template version == interface version" in {
+      val pkgs = SpeedyTestLib.typeAndCompile(
+        Map(
+          templatePkgId -> templatePkg,
+          interfacesPkgId -> interfacesPkg,
+          implementsPkgId -> implementsPkg,
+          coImplementsPkgId -> coImplementsPkg,
+        )
+      )
+
+      for ((templateId, interfaceId, contract) <- testData) {
+        val result = evaluateBeginExercise(
+          pkgs,
+          templateId,
+          Some(interfaceId),
+          contractId,
+          committers = Set(contractParty),
+          controllers = Set(contractParty),
+          getContract = Map(contractId -> contract),
+        )
+
+        inside(result) { case Right(transaction) =>
+          transaction.version shouldBe TransactionVersion.assignNodeVersion(commonVersion)
+        }
+      }
+    }
+  }
+}
+
+object TransactionVersionTest {
+  val commonVersion: LanguageVersion = LanguageVersion.default
+  val oldVersion: LanguageVersion = LanguageVersion.v1_15
+  val newVersion: LanguageVersion = LanguageVersion.v1_dev
+  val (templatePkgId, templatePkg) =
+    PackageId.assertFromString("template-pkg") -> p"""
+        module TemplateMod {
+          record @serializable Template1 = { person: Party, label: Text };
+          template (this: Template1) = {
+            precondition True;
+            signatories (Cons @Party ['template-pkg':TemplateMod:Template1 {person} this] (Nil @Party));
+            observers (Nil @Party);
+            agreement "Agreement for template Template1";
+
+            choice Destroy (self) (arg: Unit): Unit,
+              controllers (Cons @Party ['template-pkg':TemplateMod:Template1 {person} this] (Nil @Party)),
+              observers (Nil @Party)
+              to upure @Unit ();
+          };
+        }
+       """
+  val (interfacesPkgId, interfacesPkg) =
+    PackageId.assertFromString("interfaces-pkg") -> p"""
+         module InterfacesMod {
+           interface (this: Interface1) = {
+             viewtype Unit;
```

Comment:
please use a record type, as we will soon check views are records.
see #14746 for examples